File: 2022/6/solution.py
import os

def findMarker(line: str, distincts: int) -> int:
  i = 0
  n = distincts
  letters: dict[str, int] = {}
  while i < n:
    c = line[i]
    if c in letters:
      n = max(n, letters[c] + distincts + 1)
    letters[c] = i
    i += 1
  return i

if __name__ == "__main__":
  file1 = open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r')
  line = file1.read()

  print(findMarker(line, 4))
  print(findMarker(line, 14))

File: 2022/6/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := string(data)
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func findMarker(puzzleInput string, nDistinct int) int {
	letters := make(map[byte]int, nDistinct)
	for i := range puzzleInput {
		letters[puzzleInput[i]]++
		if i >= nDistinct {
			letters[puzzleInput[i-nDistinct]]--
			if letters[puzzleInput[i-nDistinct]] == 0 {
				delete(letters, puzzleInput[i-nDistinct])
			}
		}
		if len(letters) == nDistinct {
			return i + 1
		}
	}
	return -1
}

func part1(puzzleInput string) int {
	return findMarker(puzzleInput, 4)
}

func part2(puzzleInput string) int {
	return findMarker(puzzleInput, 14)
}
